eth/downloader/downloader.go

```diff
@@ -304,6 +310,13 @@ func (d *Downloader) synchronise(id string, hash common.Hash, td *big.Int, mode
 		default:
 		}
 	}
+	for empty := false; !empty; {
```

Comment:
I've been thinking about this and I think it's fine like this:

 * The "wake" channels I've mentioned above are meant for notifying some goroutines that they should wake up and continue downloading. For this reason there's logically no meaningful reason to buffer more than 1 notification.
 * On the other hand, the header channel buffers up batches of headers that should be crunched through. The buffered nature of this channel is to allow the downloader to progress without waiting for the processor to pick up the headers. Because of this the "buffer of 1 batch" is a tuning choice, but could be any other too. Moving this cleanup to among the "wake" channels would make it work only for batches of 1, and silently fail for others.